adadamp/_dist.py

```diff
@@ -590,3 +590,119 @@ def set_bs(self, bs):
         Updates batch size
         """
         self.batch_size = bs
+    
+        
+class DaskClassifierSimulator(DaskClassifierExpiriments):
+    
+    def set_sim(self, dic):
+        """
+        Sets simulation data for next epoch
+        """
+        self._sim_data = dic
+        
+    def initialize(self):
+        self._sim_data = None
+        super().initialize()
+        
+    def _get_gradients(
+        self,
+        start_idx,
+        model_opt,
+        dataset,
+        *,
+        loss,
+        batch_size,
+        client,
+        n_workers,
+        len_dataset,
+        device,
+    ):
+        """
+        Calculates the gradients at a given state. This function is
+        mainly in charge of sending the gradient calculation off to the
+        dask workers, the hard work is done in _dist.gradient()
+
+        Note:
+        - In this implementation, each worker manually grabs the data
+          and caches it themselves. This is a work around for a dask bug
+        - See the doc string of _get_fashionminst in _dist.py for more info
+
+        Parameters
+        ----------
+        start_idx : int
+            Index to start sampling at. Data indices from ``start_idx`` to
+            ``start_idx + batch_size`` will be sampled.
+
+        model_future : distributed.future (?)
+            Future object of the model to get gradients for
+
+        n : int
+            number of items in training set
+
+        idx: int
+            Current epoch, used to set up random state
+
+        batch_size: int
+            Current batch size
+
+        n_workers: int
+            Number of workers current running
+
+        """
+        # Iterate through the dataset in batches
+        # TODO: integrate with IterableDataset (this is pretty much already
+        # an IterableDataset but without vectorization)
+        idx = self.random_state_.choice(len_dataset, size=min(batch_size, len_dataset), replace=False)
+        idx.sort()
+        worker_idxs = np.array_split(idx, n_workers)
+
+        # Distribute the computation of the gradient. In practice this will
+        # mean (say) 4 GPUs to accelerate the gradient computation. Right now
+        # for ease it's a small network that doesn't need much acceleration.
+        total_time = self._sim_data["partial_fit__time"]
+        grads_per_worker = self._sim_data["grads_per_worker"]
+        batch_size = self._sim_data["partial_fit__batch_size"]
+        time = total_time / (batch_size/grads_per_worker)
```

Comment:
I have the basic simulation set up right now. I am going to try to run/test it sometime over the weekend before we meet. I have 2 main questions:
- Should the fake/simulated gradient function return an empty gradient? If so, what values / size should it be?
- Is my math on the 4 lines highlighted correct in calculating the sleep time for a worker?

Reply:
> Is my math on the 4 lines highlighted correct in calculating the sleep time for a worker?

I'd suggest a simpler method to sleep. I'd do this math:

``` python
# Time to the gradient for 128 examples
# Determined offline; this is the average time for one model update when the batch size is 128
# i.e., partial_fit__time / n_updates_in_epoch
TIME_PER_128 = 0.001

def gradient(train_idx, ...):
    assert len(train_idx) % 128 == 0, "Make sure divisible by 128"
    sleep(len(train_idx) * TIME_PER_128 / 128)
    return {name: grad for name, grad in model.named_parameters()}
```

The goal is to simulate training, and the communication is the primary simulation concern in the return value. So `gradient` should return an object of the same size as actual training because size is the primary constraint in communication. That's easiest if gradients (or random tensors of the same size) are returned.